src/state-manager/TransactionConsensus.ts

```diff
@@ -1024,6 +1094,565 @@ class TransactionConsenus {
+            nestedCountersInstance.countEvent('poqo', 'poqo-receipt-gossip: Rejecting receipt because no majority')
+            return
+          }
+          
+          queueEntry.poqoReceipt = payload
+          queueEntry.appliedReceipt2 = payload
+          queueEntry.recievedAppliedReceipt2 = payload
+          Comms.sendGossip(
+            'poqo-receipt-gossip',
+            payload,
+            null,
+            null,
+            queueEntry.transactionGroup,
+            false,
+            4,
+            payload.txid,
+            '',
+            true
+          )
+          queueEntry.hasSentFinalReceipt = true
+        } finally {
+          profilerInstance.scopedProfileSectionEnd('poqo-receipt-gossip')
+        }
+      }
+    )
+
+    const poqoDataAndReceiptBinaryHandler: Route<InternalBinaryHandler<Buffer>> = {
+      name: InternalRouteEnum.binary_poqo_data_and_receipt,
+      handler: async (payload, respond, header, sign) => {
+        const route = InternalRouteEnum.binary_poqo_data_and_receipt
+        this.profiler.scopedProfileSectionStart(route, false)
+        try {
+          const _sender = header.sender_id
+          const reqStream = getStreamWithTypeCheck(payload, TypeIdentifierEnum.cPoqoDataAndReceiptReq)
+          if (!reqStream) {
+            nestedCountersInstance.countEvent('internal', `${route}-invalid_request`)
+            return
+          }
+          const readableReq = deserializePoqoDataAndReceiptResp(reqStream)
+          // make sure we have it
+          const queueEntry = this.stateManager.transactionQueue.getQueueEntrySafe(readableReq.finalState.txid) // , payload.timestamp)
+          //It is okay to ignore this transaction if the txId is not found in the queue.
+          if (queueEntry == null) {
+            //In the past we would enqueue the TX, expecially if syncing but that has been removed.
+            //The normal mechanism of sharing TXs is good enough.
+            nestedCountersInstance.countEvent('processing', 'broadcast_finalstate_noQueueEntry')
+            return
+          }
+
+          // validate corresponding tell sender
+          if (_sender == null) {
+            /* prettier-ignore */ if (logFlags.error) this.mainLogger.error(`poqo-data-and-receipt invalid sender for txid: ${readableReq.finalState.txid}, sender: ${_sender}`)
+            return
+          }
+
+          const isValidFinalDataSender =
+            this.stateManager.transactionQueue.factValidateCorrespondingTellFinalDataSender(
+              queueEntry,
+              _sender
+            )
+          if (isValidFinalDataSender === false) {
+            /* prettier-ignore */ if (logFlags.error) this.mainLogger.error(`poqo-data-and-receipt invalid sender ${_sender} for data: ${queueEntry.acceptedTx.txId}`)
+            return
+          }
+
+          if (!queueEntry.hasSentFinalReceipt) {
+            const executionGroupNodes = new Set(queueEntry.executionGroup.map(node => node.publicKey));
```

Comment:
just curious how "expensive" this procedure. I saw it is used a lot in POQo

Reply:
it could be expensive but in the latest code (possibly changed since review time) I only see this in a two spots that happen once per TX so I think it is ok.   the execution group fortunately is limited to 128 that helps some, otherwise I would ask for this to be not declarative.

Reply:
this pattern does happen a lot so it will likely get squashed in some places in a future perf pass